tests/periph_countdown/main.c

```diff
@@ -0,0 +1,230 @@
+
+        countdown_start(i);
+        ztimer_sleep(ZTIMER_MSEC, timeout / 2);
+        countdown_cancel(i);
+
+        TEST_ASSERT_EQUAL_INT(-ECANCELED, ztimer_mutex_lock_timeout(ZTIMER_MSEC, &lock, timeout + 5));
+    }
+}
+
+static void test_countdown_long_set_cancel(void)
+{
+    mutex_t lock = MUTEX_INIT_LOCKED;
+    const unsigned timeout = 1000;
+
+    for (unsigned i = 0; i < COUNTDOWN_NUMOF; ++i) {
+        countdown_init(i, timeout / 4, COUNTDOWN_FLAG_UNIT_MS, _unlock_mutex, &lock);
+        countdown_set_long(i, timeout, COUNTDOWN_FLAG_UNIT_MS);
+
+        countdown_start(i);
+        ztimer_sleep(ZTIMER_MSEC, timeout / 2);
+
+        /* calling set must cancel the current timeout */
+        countdown_set_long(i, timeout, COUNTDOWN_FLAG_UNIT_MS);
+
+        TEST_ASSERT_EQUAL_INT(-ECANCELED, ztimer_mutex_lock_timeout(ZTIMER_MSEC, &lock, timeout + 5));
+    }
+}
+
+static void test_countdown_multi_25ms(void)
+{
+    sema_inv_t sema;
+    const unsigned timeout = 25;
+
+    sema_inv_init(&sema, COUNTDOWN_NUMOF);
+
+    for (unsigned i = 0; i < COUNTDOWN_NUMOF; ++i) {
+        TEST_ASSERT(countdown_init(i, timeout, COUNTDOWN_FLAG_UNIT_MS, _post_sema, &sema));
+        countdown_set(i, timeout, COUNTDOWN_FLAG_UNIT_MS);
+    }
+
+    ztimer_now_t before = ztimer_now(ZTIMER_MSEC);
+
+    for (unsigned i = 0; i < COUNTDOWN_NUMOF; ++i) {
+        countdown_start(i);
+    }
+
+    TEST_ASSERT_EQUAL_INT(0, ztimer_mutex_lock_timeout(ZTIMER_MSEC, &sema.lock, timeout + 5));
+    TEST_ASSERT(timeout - 5 < ztimer_now(ZTIMER_MSEC) - before);
+}
+
+static void test_countdown_multi_long_1s(void)
+{
+    sema_inv_t sema;
+    const unsigned timeout = 1000;
+
+    sema_inv_init(&sema, COUNTDOWN_NUMOF);
+
+    for (unsigned i = 0; i < COUNTDOWN_NUMOF; ++i) {
+        TEST_ASSERT(countdown_init(i, timeout / 4, COUNTDOWN_FLAG_UNIT_MS, _post_sema, &sema));
+        countdown_set_long(i, timeout, COUNTDOWN_FLAG_UNIT_MS);
+    }
+
+    ztimer_now_t before = ztimer_now(ZTIMER_MSEC);
+
+    for (unsigned i = 0; i < COUNTDOWN_NUMOF; ++i) {
+        countdown_start(i);
+    }
+
+    TEST_ASSERT_EQUAL_INT(0, ztimer_mutex_lock_timeout(ZTIMER_MSEC, &sema.lock, timeout + 5));
+    TEST_ASSERT(timeout - 5 < ztimer_now(ZTIMER_MSEC) - before);
+}
+
+static Test *tests_periph_countdown(void)
+{
+    EMB_UNIT_TESTFIXTURES(fixtures) {
+        new_TestFixture(test_countdown_25ms),
+        new_TestFixture(test_countdown_25000ns),
+        new_TestFixture(test_countdown_long_1s),
+        new_TestFixture(test_countdown_long_cancel),
+        new_TestFixture(test_countdown_long_set_cancel),
+        new_TestFixture(test_countdown_multi_25ms),
+        new_TestFixture(test_countdown_multi_long_1s),
+    };
+
+    EMB_UNIT_TESTCALLER(periph_countdown_tests, NULL, NULL, fixtures);
+
+    return (Test *)&periph_countdown_tests;
+}
+
+int main(void)
+{
+    TESTS_START();
+    TESTS_RUN(tests_periph_countdown());
+    TESTS_END();
+
+    puts("\n--- Benchmark ---");
```

Comment:
do you mind using sys/benchmark here?

Reply:
Ok, although that's almost a bit too verbose

Reply:
but it'll support turo soon, which is parsed by CI. small steps to tracking regressions on benchmark values, or creating overviews of board performance, automaticalls...

So thanks! :)